umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj

```diff
@@ -52,12 +52,42 @@
                                             gmd-id "gmd:MD_Identifier/gmd:code/gmx:Anchor"))
                              explanation (parse-explanation
                                           (value-of gmd-id
-                                                    "gmd:MD_Identifier/gmd:description/gco:CharacterString"))]]
+                                                    "gmd:MD_Identifier/gmd:description/gco:CharacterString"))
+                             attrs (get (first (select gmd-id "gmd:MD_Identifier/gmd:code")) :attrs)
+                             missing-reason (when attrs
+                                              (cond
+                                                (string/includes? attrs "inapplicable") "Not Applicable"
+                                                (string/includes? attrs "unknown") "Unknown"
+                                                :else nil))]]
                    (util/remove-nil-keys
-                    {:DOI doi-value
-                     :Authority (or (value-of gmd-id orgname-path)
-                                    (value-of gmd-id indname-path))
-                     :MissingReason (when-not (seq doi-value)
-                                      "Not Applicable")
-                     :Explanation explanation}))]
-    (first doi-list)))
+                     (if doi-value
+                       {:DOI doi-value
+                        :Authority (or (value-of gmd-id orgname-path)
+                                       (value-of gmd-id indname-path))}
+                       {:MissingReason (when missing-reason
+                                         missing-reason)
+
+                        :Explanation explanation})))]
+    (if (first doi-list)
+      (first doi-list)
+      {:MissingReason "Unknown"
+       :Explanation "It is unknown if this record has a DOI."})))
+
+(defn parse-associated-dois
+  "Parse out the associated DOIs from the ISO MENDS/SMAP document."
+  [doc associated-doi-xpath]
+  (let [assocs
+         (seq
+           (for [assoc (select doc associated-doi-xpath)
+                 :let [assoc-type (value-of assoc "gmd:associationType/gmd:DS_AssociationTypeCode")
+                       assoc-type (when assoc-type
+                                    (string/trim assoc-type))
+                       doi (iso-util/char-string-value assoc "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier/gmd:code")]
+                 :when (and doi (= "associatedDOI" assoc-type))]
+             {:DOI doi
+              :Title (iso-util/char-string-value assoc "gmd:aggregateDataSetName/gmd:CI_Citation/gmd:title")
+              :Authority (iso-util/char-string-value assoc (str "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier"
+                                                                "/gmd:authority/gmd:CI_Citation/gmd:citedResponsibleParty"
+                                                                "/gmd:CI_ResponsibleParty/gmd:organisationName"))}))]
```

Comment:
This feels like a lot for a `let`, but the function itself does so little that I'm not sure if it's worth putting this into its own function. If it's possible to clean it up that would be awesome, but I don't have any suggestions that I like, so I won't put up a fuss 🙂

Reply:
Hehe,  I don't think it is worth it either, so I will leave it.

Reply:
I think the simplest refactor to improve readability would be something like this
```
(defn assemble-doi
  [doi]
  {:DOI doi
   :Title (iso-util/char-string-value assoc "gmd:aggregateDataSetName/gmd:CI_Citation/gmd:title")
   :Authority (iso-util/char-string-value assoc (str "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier"
                                                     "/gmd:authority/gmd:CI_Citation/gmd:citedResponsibleParty"
                                                     "/gmd:CI_ResponsibleParty/gmd:organisationName"))})
(defn parse-associated-dois
  [doc associated-doi-xpath]
  (let [assocs
        (seq
         (for [assoc (select doc associated-doi-xpath)
               :let [assoc-type (value-of assoc "gmd:associationType/gmd:DS_AssociationTypeCode")
                     assoc-type (when assoc-type
                                  (string/trim assoc-type))
                     doi (iso-util/char-string-value assoc "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier/gmd:code")]
               :when (and doi (= "associatedDOI" assoc-type))]
           (asseble-doi doi)))]))
           ```